### PR TITLE
Feature/diffae

### PR DESCRIPTION
**Pull request by benjijamorris**
## What does this PR do?

Add diffusion autoencoder with callback for doing latent walks during training
add concatenation resolver

## Before submitting

- [x] Did you make sure **title is self-explanatory** and **the description concisely explains the PR**?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **test your PR locally** with `pytest` command?
- [x] Did you **run pre-commit hooks** with `pre-commit run -a` command?

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: configs/data/im2im/diffae.yaml [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    _target_: monai.transforms.Compose
+    transforms:
+      # channels are [blank, membrane,blank, structure, blank, nuclear dye, brightfield ]
+      - _target_: monai.transforms.LoadImaged
+        keys: ${source_col}
+        reader:
+          - _target_: cyto_dl.image.io.MonaiBioReader
+            # NOTE: eval is used so only the experiment file is required to change for beginning users. This is not recommended when creating your own configs.
+            dimension_order_out: ${eval:'"CZYX" if ${spatial_dims}==3 else "CYX"'}
+            C: 5
```
Comment (ritvikvasan) [inline configs/data/im2im/diffae.yaml:24]: is this picking the fifth channel?

File: configs/data/im2im/diffae.yaml [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        keys:
+          - ${source_col}
+        roi_size: ${data._aux.patch_shape}
+        num_samples: 1
+        random_size: False
+
+  test:
+    _target_: monai.transforms.Compose
+    transforms:
+      # channels are [blank, membrane,blank, structure, blank, nuclear dye, brightfield ]
```
Comment (ritvikvasan) [inline configs/data/im2im/diffae.yaml:49]: what are blanks?

File: configs/experiment/im2im/diffae.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+# only source_col is needed for masked autoencoder
+source_col: raw
+spatial_dims: 3
+raw_im_channels: 1
+
+trainer:
+  max_epochs: 100
+  gradient_clip_val: 10
+  # 32-bit precision is only used for cpu testing, this can be changed if a gpu is used
+  precision: 32
```
Comment (ritvikvasan) [inline configs/experiment/im2im/diffae.yaml:29]: this is run on a cpu?

File: cyto_dl/models/im2im/diffusion_autoencoder.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+        _DEFAULT_METRICS = {
+            "train/loss": MeanMetric(),
+            "val/loss": MeanMetric(),
+            "test/loss": MeanMetric(),
+        }
+
+        metrics = base_kwargs.pop("metrics", _DEFAULT_METRICS)
+        super().__init__(metrics=metrics, **base_kwargs)
+        self.diffusion_key = diffusion_key or condition_key
+        self.automatic_optimization = True
```
Comment (ritvikvasan) [inline cyto_dl/models/im2im/diffusion_autoencoder.py:82]: doesn't this default to True?

File: cyto_dl/models/im2im/diffusion_autoencoder.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+            sample, intermediates = self.inferer.sample(
+                input_noise=noise,
+                diffusion_model=self.autoencoder,
+                scheduler=self.scheduler,
+                save_intermediates=True,
+                conditioning=cond,
+                verbose=False,
+                intermediate_steps=1,
+            )
+        # during training, final image is all-nan, try to return last non-nan image
```
Comment (ritvikvasan) [inline cyto_dl/models/im2im/diffusion_autoencoder.py:135]: why is this?

File: cyto_dl/models/im2im/diffusion_autoencoder.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+        noise = torch.randn_like(cond_img, device=self.device)
+        sample = self._generate_image(noise, cond)
+
+        for img, name in zip([cond_img, diff_img, sample], ["cond", "diff", "recon"]):
+            OmeTiffWriter.save(
+                uri=f"{self.hparams.save_dir}/{self.trainer.current_epoch}_{stage}_{name}.tiff",
+                data=detach(img).astype(float),
+            )
+
+    def _to_tensor(self, batch):
```
Comment (ritvikvasan) [inline cyto_dl/models/im2im/diffusion_autoencoder.py:155]: this seems like a general utility

File: cyto_dl/models/im2im/diffusion_autoencoder.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+        diffusion_loss = self.loss(noise, noise_pred)
+
+        return {"loss": diffusion_loss}, latent, None
+
+    def validation_step(self, batch, batch_idx):
+        loss, preds, targets = self.model_step("val", batch, batch_idx)
+        self.compute_metrics(loss, preds, targets, "val")
+        return loss, preds
+
+    def latent_walk(self, cond, save_name):
```
Comment (ritvikvasan) [inline cyto_dl/models/im2im/diffusion_autoencoder.py:178]: is this also a callback?

File: cyto_dl/models/im2im/diffusion_autoencoder.py [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+                noise = torch.stack([torch.randn([1] + self.hparams.image_shape, device=self.device)] * cond.shape[0])
+                sample = self._generate_image(noise, cond.unsqueeze(2))
+                recon = sample if recon is None else recon + sample
+            # average reconstruction across samples
+            recon /= self.hparams.n_noise_samples
+        OmeTiffWriter.save(
+            uri=f"{self.hparams.save_dir}/{save_name}.tiff", data=detach(recon).astype(float)
+        )
+
+    def _generate_and_re_embed(self, z, batch_idx):
```
Comment (ritvikvasan) [inline cyto_dl/models/im2im/diffusion_autoencoder.py:193]: cool!! is this loss not a part of training as well?

**Review by ritvikvasan — CHANGES_REQUESTED** [review]
Super cool! This PR seems to be missing

`cyto_dl.models.im2im.utils.inferers.EmbeddingPatchInferer`
`cyto_dl.models.im2im.utils.inferers.EmbeddingPatchMerger`

for the spatial_inferer

File: configs/data/im2im/diffae.yaml (hunk 1 above)
Comment (benjijamorris) [reply configs/data/im2im/diffae.yaml:24]: yes, I use variance data for all the tests

File: configs/data/im2im/diffae.yaml (hunk 2 above)
Comment (benjijamorris) [reply configs/data/im2im/diffae.yaml:49]: just channels in variance data without a structure imaged

File: configs/experiment/im2im/diffae.yaml (hunk 3 above)
Comment (benjijamorris) [reply configs/experiment/im2im/diffae.yaml:29]: yes, all tests are run on cpu

File: cyto_dl/models/im2im/diffusion_autoencoder.py (hunk 4 above)
Comment (benjijamorris) [reply cyto_dl/models/im2im/diffusion_autoencoder.py:82]: updated!

File: cyto_dl/models/im2im/diffusion_autoencoder.py (hunk 5 above)
Comment (benjijamorris) [reply cyto_dl/models/im2im/diffusion_autoencoder.py:135]: not sure.... it goes away during inference

**Review by ritvikvasan — CHANGES_REQUESTED** [review]
Also missing - 

```
DDIMScheduler
DiffusionInferer
```

File: cyto_dl/models/im2im/diffusion_autoencoder.py (hunk 7 above)
Comment (benjijamorris) [reply cyto_dl/models/im2im/diffusion_autoencoder.py:178]: yes, the callback calls this function. I have it split out so you can generate a sequence of images from a batch of latent variables. I'll rename this to make it more clear.

File: cyto_dl/models/im2im/diffusion_autoencoder.py (hunk 8 above)
Comment (benjijamorris) [reply cyto_dl/models/im2im/diffusion_autoencoder.py:193]: This was more a validation to see if the semantic latent was actually being used, maybe not worth including in the PR.

**Comment by benjijamorris** [comment]
Those two should be imported from monai and monai-generative

**Comment by ritvikvasan** [comment]
got it. Is the toml file updated to include this dependency?

**Comment by benjijamorris** [comment]
yeah! #

**Comment by benjijamorris** [comment]
yeah, I have a lot of random changes and I wanted to merge them into that branch, test them there, and then merge to main without making one massive pr

**Review by ritvikvasan — APPROVED** [review]

**Comment by ritvikvasan** [comment]
great! LGTM
